win32ss/gdi/ntgdi/region.c

```diff
@@ -2149,9 +2149,9 @@ REGION_bXformRgn(
             /* Loop all rects in the region */
             for (i = 0; i < prgn->rdh.nCount - 1; i++)
             {
-                for (j = i; i < prgn->rdh.nCount; i++)
+                for (j = i; j < prgn->rdh.nCount; j++)
                 {
-                    NT_ASSERT(prgn->Buffer[i].top < prgn->Buffer[i].bottom);
+                    NT_ASSERT(prgn->Buffer[i].top <= prgn->Buffer[i].bottom);
                     NT_ASSERT(prgn->Buffer[j].top >= prgn->Buffer[i].top);
```

Comment:
The "larger or equal" operator is transitive which means if `a[i + 1] >= a[i]` and `a[i + 2] >= a[i + 1]`, then `a[i + 2] >= a[i]`.
... which in plain terms means the nested loop is completely unnecessary. It's enough to `NT_ASSERT(prgn->Buffer[i + 1].top >= prgn->Buffer[i].top)` in the outer loop and we'll have sufficiently checked the invariant.

Would you mind making that extra simplification?

Reply:
Note that, though getting rid of the (broken) inner loop looks like what was initially intended, it actually "reverses" this 2nd check. As in a bug that would have never been triggered?
Please, double-check PR code is as wanted now.

Reply:
The invariant is: the `top` value for higher indices is higher (or equal). Both the old j/i check and the new +1/+0 check should validate the same thing, or am I missing something?

Reply:
The old check was actually
`NT_ASSERT(prgn->Buffer[ 0 ].top >= prgn->Buffer[ 0 ... (nCount - 1) ].top);`

Reply:
First let's define the intent of two things what these assertions were intended to detect.
First was the fact that by comparing the top and the bottom of each rectangle, in no case would the top be more than the bottom because this "is" a well-ordered rectangle.
Second we want to see if they are sorted such that their tops are sorted in lowest to highest ordering, but may be equal as well.
So if T(x) represents the top value and x represents the rectangle number 0 through whatever, then we want the following to be true.
T(0) <= T(1) <= T(2) <= T(3) , etc.
Now if we have numbers as follows 1,2,3,4, these meet our criteria.
If we check each pair (0,1), (1,2), (2,3), and (3,4) which is our new way these meet our needs.
We could also check (0,1), (0,2), (0,3), (0.4), (1,2), (1,3), (1,4), (2,3), (2,4) and (3,4) which is the old way and these would be good as well.
If we agree that what we want from the assertions is to know if our previous well-ordering algorithm worked, then if we test T(x) <= T(x+1)  for x = 0,1,2,... count-1 and all of these are good, we don't have to check all combinations of the two. For example if our number series were 8,1,2,3 then we would fail on the first four pair tests with the old code, but only fail once on our present code.
If we have ONE error we need to fix it and having it reported multiple times is of little benefit as far as I am concerned, so I believe that our new code accomplishes this task well.

Reply:
I am not sure what your point is.
1st check: what about the new slight _regression_?
2nd check: old code is **not** currently doing what you write...

Reply:
We've both said we can't follow your logic, so you have to get more specific, not less.

In the old code for the second check, j>=i, so now we replace j with i+1, since the transitivity of the comparison makes it enough to compare the next item rather than all following ones.

In the first check, I understand your point about not catching the last item (but it's not a regression), I understand your point about MakeWellOrdered (but disagree; most asserts should never fire, that's the point), and I understand your point about a future fix (but it's irrelevant to this change); so I fail to find a slight, or otherwise any, regression.

Reply:
> can't follow your logic, so you have to get more specific

I tried with text and code, not sure what else I can do. Actually run the double-loop for yourself?
It is not logic of mine only: double-check [CORE-12204 Katayama's patch](https://jira.reactos.org/secure/attachment/41267/region.c.patch).

> second check, j>=i

No! `j == 0`...

> not catching the last item (but it's not a regression)

I don't know how you call it then.
Anyway, why is it correct not to check it anymore?

Reply:
So what you're saying is that you're comparing to the `for (j = i; i < prgn->rdh.nCount; i++)` loop without fixing the obvious typos where the code says `i` instead of `j`. Okay. That's not useful.

A regression is when something used to be correct but now is not. The assert did not check the last item before this change, and it does not check it after this change. No change in behavior, no regression.

Reply:
> That's not useful.

Hum, indeed, I mean "old == currently committed". It seems Doug understood it that way too.
Comparing PR behavior/result seems interesting to me...

> The assert did not check the last item before this change

I disagree with that: currently committed code does check all `i` values, from `0` to `nCount - 1`

Reply:
Your original comment was:
> Note that, though getting rid of the (broken) inner loop looks like what was initially intended, it actually "reverses" this 2nd check

That indicates to me that you think something was correct after the typo fix but behavior was different after removing the inner loop. If the typo fix already caused your problem, you need to say so, and not say the opposite. As I said: be more precise!

> I disagree with you about that: currently committed code does check all i values, from 0 to nCount - 1

Okay. Again, comparisons to the code with the obvious typo in it are NOT USEFUL.

Reply:
> be more precise

I accept the blame if my sentence was not explicit enough.
Nonetheless, you probably need to be more precise too: you started using the word "old", now you refer to a typo fix that was not committed, then how can the latter be "old"?
"Lost in translation"?

> NOT USEFUL

That totally confuses me: how is a regression defined if not by comparing "states" before and after a commit?